streams/readable-streams/async-iterator.js

```diff
@@ -0,0 +1,145 @@
+'use strict';
+
+if (self.importScripts) {
+  self.importScripts('../resources/rs-utils.js');
+  self.importScripts('/resources/testharness.js');
+  self.importScripts('../resources/recording-streams.js');
+}
+
+test(() => {
+  assert_equals(ReadableStream.prototype[Symbol.asyncIterator], ReadableStream.prototype.getIterator);
+}, '@@asyncIterator() method is === to getIterator() method');
+
+promise_test(async () => {
+  const s = recordingReadableStream({
+    start(c) {
+      c.enqueue(1);
+      c.enqueue(2);
+      c.enqueue(3);
+      c.close();
+    },
+  });
+
+  const chunks = [];
+  for await (const chunk of s) {
+    chunks.push(chunk);
+  }
+  assert_array_equals(chunks, [1, 2, 3]);
+}, 'async iterator push source');
+
+promise_test(async () => {
+  let i = 1;
+  const s = recordingReadableStream({
+    pull(c) {
+      c.enqueue(i);
+      if (i >= 3) {
+        c.close();
+      }
+      i += 1;
+    },
+  });
+
+  const chunks = [];
+  for await (const chunk of s) {
+    chunks.push(chunk);
+  }
+  assert_array_equals(chunks, [1, 2, 3]);
+}, 'async iterator pull source');
+
+promise_test(async () => {
+  const s = new ReadableStream({
+    start(c) {
+      c.error('e');
+    },
+  });
+
+  try {
+    for await (const chunk of s) {}
+    assert(false);
+  } catch (e) {
+    assert_equals(e, 'e');
+  }
+}, 'Async-iterating an errored stream throws');
+
+promise_test(async () => {
+  const s = new ReadableStream({
+    start(s) {
+      s.close();
+    }
+  });
+
+  for await (const chunk of s) {
+    assert(false);
+  }
+}, 'Async-iterating a closed stream never executes the loop body, but works fine');
+
+promise_test(async () => {
+
+}, 'Async-iterating an empty but not closed/errored stream never executes the loop body and stalls the async function');
+
+promise_test(async () => {
+  const test = async (type, preventCancel) => {
+    const s = recordingReadableStream({
+      start(c) {
+        c.enqueue(0);
+      }
+    });
+
+    await (async () => {
+      for await (const c of s.getIterator({ preventCancel })) {
+        if (type === 'throw') {
+          throw 'e';
+        } else if (type === 'break') {
+          break;
+        } else if (type === 'return') {
+          return;
+        }
+      }
+    })().catch(() => 0);
+
+    assert_equals(s.locked, preventCancel);
```

Comment:
It's easier to debug assert failures if they have a description. In this case, I'd use something like ``` `s.locked should equal preventCancel when type = ${type} and preventCancel = ${preventCancel}` ```.

In this case, the assertion is wrong because `s.locked` is always false after the loop terminates. The easiest way to detect if `preventCancel` worked is to look at `s.events`. Something like this should work:
```javascript
if (preventCancel) {
  assert_array_equals(s.events, [], `cancel() should not be called when type = '${type}' and preventCancel is true`);
} else {
  assert_array_equals(s.events, ['cancel', undefined], `cancel() should be called when type = '${type}' and preventCancel is false`);
}
```